### PR TITLE
[jit] Change interpreter/fuser to work on Variables only

### PR DESCRIPTION
**Pull request by zdevito**
* this removes the flag controlling whether the interpreter works on variables.
* now the interpreter _always_ works on variables
* constants in the IR are still _always_ non-variables, and an assert was added to ensure this.
* as_tensor was split into as_variable and as_tensor since it is sometimes used
  to construct constants in the IR
* I tried changing the IR to also always use variables but that change was much more
  cross cutting and fragile and I never got it working

Note: This change is blocking the translation work so it should be given high priority.

### CODE REVIEW

**Review by apaszke — APPROVED** [review]
cc @goldsborough for C++ API changes

I'm really concerned about the growing complexity of GraphExecutor. It's already pretty hard to follow, and this patch makes it even worse. We should stop dumping hotfixes into the JIT and actually start refactoring it properly.

File: torch/csrc/torch_api.h [hunk 1]
```diff
@@ -0,0 +1,30 @@
+#pragma once
```
[the remark this answers is not in the document]
Comment (goldsborough) [reply torch/csrc/torch_api.h]: We can’t do this unless we rename it to torch.h in setup.py. The header must not change for users, otherwise all C++ extensions will break 2 weeks after releasing them

File: torch/csrc/jit/attributes.h [hunk 2]
```diff
@@ -167,13 +167,30 @@ struct Attributes {
...
-  CREATE_ACCESSOR(Tensor,t)
-  CREATE_ACCESSOR(Tensors,ts)
   CREATE_ACCESSOR(Graph,g)
   CREATE_ACCESSOR(Graphs,gs)
 
   #undef CREATE_ACCESSOR
 
+  // does not use CREATE_ACCESSOR because we need additional asserts
+  Derived* t_(Symbol name, TensorAttr::ConstructorType v) {
+    JIT_ASSERT(!v.defined() || !v.is_variable_or_undefined());
```
[the remark this answers is not in the document]
Comment (jamesr66a) [reply torch/csrc/jit/attributes.h]: @apaszke I think you're missing the outer negation after applying DeMorgan's. It should look like:

`JIT_ASSERT(!(v.defined() && v.is_variable_or_undefined()))`

Which I think is actually less clear than what we have here

File: torch/csrc/jit/graph_executor.cpp [hunk 3]
```diff
@@ -430,6 +443,12 @@ struct GraphExecutorImpl {
   // symbolically differentiable.
   bool symbolically_differentiable;
 
+  // some ops, including python operations, can intorduce requires_grad=True
+  // variables even though no inputs to this graph are availiable, if
+  // the graph includes those operators then needGradient must be true
+  // regardles of input state.
+  bool may_introduce_gradient;
```
[the remark this answers is not in the document]
Comment (zdevito) [reply torch/csrc/jit/graph_executor.cpp]: I agree, these things are getting added because of actual bugs.  This one is because we changed an invariant we used to have: that the python code we call is functional. Now python code we call may be a call to a module forward with some hidden module weights that do require grad. This is pretty common when the input tensor does not requires grad but all the weights do.

**Comment by jamesr66a** [comment]
Are there tests exercising this change? Can we at least add `requires_grad=True` to the input in `test_trace_size`?

File: torch/csrc/torch_api.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (zdevito) [reply torch/csrc/torch_api.h]: This change was just to separate the python/non-python parts of this header so that it can be included in no-python files. I will rename torch_api.h to something more descriptive.
